packages/clay-shared/src/useFocusManagement.ts

```diff
@@ -0,0 +1,224 @@
+			const index = focusManagerRef.current.findIndex(
+				el => el.key === element.key
+			);
+			element = focusManagerRef.current[index - 1];
+		}
+
+		return element;
+	};
+
+	const handleNextElementOutsideScope = (event: KeyboardEvent) => {
+		if (event.keyCode === TAB_KEY_CODE && event.shiftKey) {
+			event.preventDefault();
+			lastPositionRef.current = focusManagerRef.current.length;
+			moveFocusInScope(true);
+		}
+	};
+
+	const moveFocusInScope = (backwards: boolean = false) => {
+		const size = focusManagerRef.current.length - 1;
+		const activeNode = document.activeElement;
+		const firstPosition = lastPositionRef.current === 0;
+		const endPosition =
+			lastPositionRef.current !== null &&
+			size === lastPositionRef.current;
+
+		// Get the parent elements or focusable children of the
+		// non-portal element to move focus.
+		if (
+			!loop &&
+			(((lastPositionRef.current === null || firstPosition) &&
+				backwards) ||
+				(endPosition && !backwards))
+		) {
+			const elementOutsidePortal = getElementOutsidePortal(
+				focusManagerRef.current[
+					firstPosition || lastPositionRef.current === null ? 0 : size
+				]
+			);
+			const elementOutsideScope = getFocusableElementOutsideScope(
+				elementOutsidePortal.value,
+				backwards
+			);
+
+			// If can not find some element that is focusable in the DOM,
+			// continue with the focus loop.
+			if (
+				elementOutsideScope &&
+				focusManagerRef.current.findIndex(
+					el => el.value === elementOutsideScope
+				) === -1
+			) {
+				elementOutsideScope.focus();
+				lastPositionRef.current = null;
+
+				// Control the tab + shift keydown of the element next to the scopo to
+				// return the focus to the last scopo element, this is necessary for
+				// when have React Portal.
+				if (!backwards) {
+					if (
+						nextElementOutsideScopeRef.current !== null &&
+						nextElementOutsideScopeRef.current !==
+							elementOutsideScope
+					) {
+						nextElementOutsideScopeRef.current.removeEventListener(
+							'keydown',
+							handleNextElementOutsideScope
+						);
+					}
+
+					elementOutsideScope.addEventListener(
+						'keydown',
+						handleNextElementOutsideScope
+					);
+					nextElementOutsideScopeRef.current = elementOutsideScope;
+				}
+
+				return null;
+			}
+		}
+
+		if (
+			lastPositionRef.current === null ||
+			(firstPosition && backwards) ||
+			(endPosition && !backwards)
+		) {
+			if (backwards) {
+				lastPositionRef.current = size;
+			} else {
+				lastPositionRef.current =
+					activeNode === focusManagerRef.current[0].value ? 1 : 0;
+			}
+		} else {
+			lastPositionRef.current = backwards
+				? lastPositionRef.current - 1
+				: lastPositionRef.current + 1;
+		}
+
+		const {value} = focusManagerRef.current[lastPositionRef.current];
+
+		// Hack to check if element is visible
+		if (!value.offsetParent) {
+			moveFocusInScope(backwards);
+		} else {
+			value.focus();
+		}
+	};
+
+	useEffect(
+		() => () => {
```

Comment:
just a thought, I was a little confused when I first saw this, but then I realized you are wanting this to run when it "unmounts". It might help a bit if we avoid implicit returns for useEffect, just to make it really clear its an unmount function.

Reply:
Oh yeah, this may sound messy, you're right I'll avoid this. I've seen in some cases people creating hooks to avoid these kinds of thing, like `useUnmount` and `useMount`.